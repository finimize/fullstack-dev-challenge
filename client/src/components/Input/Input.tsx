import React, { FC } from 'react'
import { Box, Input as ChakraInput, FormLabel, Text } from '@chakra-ui/react'

type InputProps = {
    value?: string
    label: string
    placeholder?: string
    error?: string
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void
}

export const Input: FC<InputProps> = ({ label, placeholder, error, onChange, value }) => (
    <Box padding='2' maxWidth='96' marginBottom='2'>
        <FormLabel fontSize='sm' marginBottom='2' color='grey5' fontWeight='500'>
            {label}
        </FormLabel>
        <ChakraInput
            placeholder={placeholder}
            isInvalid={!!error}
            value={value}
            onChange={onChange}
        />
    </Box>
)
