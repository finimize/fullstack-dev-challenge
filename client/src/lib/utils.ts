import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
