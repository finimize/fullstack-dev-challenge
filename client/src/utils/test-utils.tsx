import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
