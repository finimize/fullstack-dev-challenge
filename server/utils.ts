const compoundedInterest = (interestRate: number, timeInMonths: number) :number => {
    return Math.pow((interestRate + 1), timeInMonths);
};

const percentageToDecimal = (percentValue: number) :number => percentValue/100;

export const monthlyCompoundInterest = (monthlyDeposit: number, monthlyInterest: number, timeInMonths: number) :number =>  {
    const interestRate = percentageToDecimal(monthlyInterest);
    const interestOverMonths = compoundedInterest(interestRate, timeInMonths);
    const monthlyCompound = (monthlyDeposit *((interestOverMonths-1)/interestRate));
    return Math.round(monthlyCompound);
}

export const principleCompoundInterest = (initialDeposit: number,
                                          monthlyInterest: number,
                                          timeInMonths: number) : number => {
    const interestOverMonths = compoundedInterest(percentageToDecimal(monthlyInterest), timeInMonths);
    return Math.round(initialDeposit * interestOverMonths);
}

export const totalSavingsOverTime = (initialDeposit: number,
                                    monthlyDeposit: number,
                                    monthlyInterest: number,
                                    timeInMonths: number) :number => {
    const principle = principleCompoundInterest(initialDeposit, monthlyInterest, timeInMonths);
    const  monthly = monthlyCompoundInterest(monthlyDeposit, monthlyInterest, timeInMonths);
    return monthly+principle;
}

const toRange = (value: number) => {
    let array = [];
    for(let i =1;i<=value;i++){
        array.push(i);
    }
    return array;
}
export const toYearsFromNow = (years: number) :Array<string> => {
    let yearsArray = toRange(years);
    return yearsArray.map(number => {
        let d = new Date();
        let currentYear = d.getFullYear();
        let month = d.getMonth();
        let day = d.getDate();
        let c = new Date(currentYear + number, month, day);
        return c.getFullYear().toString();
    })
}
export const yearlySavingsOverTime = (initialDeposit: number,
                                        monthlyDeposit: number,
                                         monthlyInterest: number,
                                         timeInYears: number) : Array<number> => {
        const arr = toRange(timeInYears);
        const principleCompound = principleCompoundInterest(initialDeposit,monthlyInterest,timeInYears*12);
        return arr.map(number => monthlyCompoundInterest(monthlyDeposit,monthlyInterest,number*12) + principleCompound);
}
