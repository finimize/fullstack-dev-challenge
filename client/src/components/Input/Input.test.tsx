import React from 'react'
import { cleanup } from '@testing-library/react'
import { customRender } from '../../utils/test-utils'
import { Input } from './Input'

afterEach(cleanup)

describe('component/CalculatorMode', () => {
    it('renders Calculator Mode component', () => {
        const { render } = customRender(<Input label='test' />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })

    it('renders Calculator Mode component', () => {
        const { render } = customRender(<Input label='number test' isNumberInput />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })
})
