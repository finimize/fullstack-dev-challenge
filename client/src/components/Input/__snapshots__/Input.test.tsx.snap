// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/CalculatorMode renders Calculator Mode component 1`] = `
<div>
   
  <div
    class="css-1d1yckv"
  >
    <label
      class="chakra-form__label css-wuh3rg"
    >
      test
    </label>
    <input
      class="chakra-input css-1ttok2p"
      data-testid="input-test"
      value=""
    />
    <p
      class="chakra-text css-80ht2y"
    />
  </div>
   
   
</div>
`;

exports[`component/CalculatorMode renders Calculator Mode component 2`] = `
<div>
   
  <div
    class="css-1d1yckv"
  >
    <label
      class="chakra-form__label css-wuh3rg"
    >
      number test
    </label>
    <div
      class="css-ylxpfg"
    >
      <input
        aria-invalid="false"
        aria-valuemax="9007199254740991"
        aria-valuemin="0"
        autocomplete="off"
        autocorrect="off"
        class="css-1ttok2p"
        inputmode="decimal"
        pattern="[0-9]*(.[0-9]+)?"
        role="spinbutton"
        type="text"
        value=""
      />
      <div
        aria-hidden="true"
        class="css-1rscer7"
      >
        <div
          class="css-9pjy17"
          role="button"
          tabindex="-1"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="css-9pjy17"
          role="button"
          tabindex="-1"
        >
          <svg
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
    <p
      class="chakra-text css-80ht2y"
    />
  </div>
   
   
</div>
`;
